langchain/vectorstores/qdrant.py

```diff
@@ -181,7 +180,7 @@ def add_texts(
             # Take the corresponding metadata for each text in a batch
             batch_metadatas = list(islice(metadatas_iterator, batch_size)) or None
 
-            batch_ids = [md5(text.encode("utf-8")).hexdigest() for text in batch_texts]
+            batch_ids = [uuid.uuid4().hex for _ in batch_texts]
```

Comment:
I may understand it wrong, but this id is used for `upsert` operation not for `insert`.
That means it should update the existing record. That's why id was previously generated from the text.
Now we cannot update the existing records because id is irrelevant to the text.
Probably to handle duplicate texts you can use the original code but add metadata to the text, something like
`md5((text+metadata_str).encode("utf-8")).hexdigest()`
when `metadata_str` is the metadata converted to a string.

Reply:
imho managing id's should be handled outside this function. we could add an optional ids param that user can pass in which would allow updating existing texts, if that's not provided then we generate new uuids

```python
def add_texts(..., ids: Optional[Sequence[str]] = None):
  _ids = ids or (uuid.uuid4().hex for _ in texts)
  batch_ids = list(islice(_ids, batch_size) )
  ...
```
cc @eyurtsev

Reply:
do you want to add ids kwarg and docstring in this pr or sep one? fine either way

Reply:
@ dev2049 It's easier to add it here. It's done and covered with some more integration tests. I used that chance to restructure the code and put the private methods at the bottom. I would appreciate your another look.